plugins/modules/s3_object_info.py

```diff
@@ -0,0 +1,658 @@
+        elif key == 'object_lock_configuration':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_lock_configuration(connection, bucket_name)
+        elif key == 'object_retention':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_retention(connection, bucket_name, object_name)
+        elif key == 'object_tagging':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_tagging(connection, bucket_name, object_name)
+
+    return all_facts
+
+
+def get_object(connection, bucket_name, object_name):
+    params = {}
+    params['Bucket'] = bucket_name
+    params['Key'] = object_name
+
+    result = {}
+    object_info = {}
+
+    try:
+        object_info = connection.get_object(**params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        pass
+
+    if len(object_info) != 0:
+        # Remove Body, ResponseMetadata from object_info, convert to snake_case
+        del(object_info['Body'])
+        del(object_info['ResponseMetadata'])
+        object_info = camel_dict_to_snake_dict(object_info)
+
+    result['object_data'] = object_info
+
+    return result
+
+
+@AWSRetry.jittered_backoff(retries=10)
+def _list_bucket_objects(connection, **params):
+    paginator = connection.get_paginator('list_objects')
+    return paginator.paginate(**params).build_full_result()
+
+
+def list_bucket_objects(connection, module, bucket_name):
+    params = {}
+    params['Bucket'] = bucket_name
+
+    result = []
+    list_objects_response = {}
+
+    try:
+        list_objects_response = _list_bucket_objects(connection, **params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        pass
+
+    if len(list_objects_response) != 0:
+        # convert to snake_case
+        for response_list_item in list_objects_response['Contents']:
+            result.append(response_list_item['Key'])
+
+    module.exit_json(s3_keys=result)
+
+
+def main():
+
+    argument_spec = dict(
+        object_details=dict(type='dict', options=dict(
+            object_acl=dict(type='bool', default=False),
+            object_legal_hold=dict(type='bool', default=False),
+            object_lock_configuration=dict(type='bool', default=False),
+            object_retention=dict(type='bool', default=False),
+            object_tagging=dict(type='bool', default=False),
+            object_attributes=dict(type='bool', default=False),
+            attributes_list=dict(type='list', elements='str', choices=['ETag', 'Checksum', 'ObjectParts', 'StorageClass', 'ObjectSize']),
+        )),
+        bucket_name=dict(required=True, type='str'),
+        object_name=dict(type='str'),
+    )
+
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+    )
+
+    try:
+        connection = module.client('s3', retry_decorator=AWSRetry.jittered_backoff())
```

Comment:
I've been thinking about this.  One problem with using module.client('s3'...) for s3_object_info...

s3_object currently has a number of 'forbidden' states (eg. list), and supports not just AWS, but supports other 'S3-compatible' services such as 'FakeS3' and Ceph/RDB.  Adding s3_object_info which only supports AWS is likely to cause us problems further down the line.

Reply:
I have mixed feeling about this.  We have a lot of modules that attempt to support various reimplementations of the S3 API, but we have no ability to test/debug these other services.  Probably suggests that someday this is going to bite us.  For the modules that already implement handling for ceph, DO, etc we shouldn't disable that support lightly but I'm not convinced we should always try to support anything other than AWS for new modules/features.

Reply:
The trouble is that to drop the 'forbidden' states from s3_object we need to make sure that s3_object_info supports it.  If s3_object_info doesn't support the same APIs then dropping support for 'list' et al in s3_object means completely dropping support.  For any other module I agree, it's lucky happenstance if it works.  For S3 because we've explicitly said we support them I feel like we're boxed in

Reply:
Ah yeah, I forgot we were doing that here.  You're right then, but I don't like it.  :smiley:

Reply:
@jillr @tremble I have tried to [add the code here](https://github.com/ansible-collections/amazon.aws/pull/977/commits/7ce7b0e1e678190d3539cfbc44a7e6a3925c1f76) that will support `list_keys` for 's3-compatible' services but would be great help to get guidance on this with my limited experience with such services.
Also not entirely sure how to test the code here? do we have capacity to test the changes with instances in those services?

Reply:
Just realized that s3_object has had recent changes in param specifications with https://github.com/ansible-collections/amazon.aws/pull/994/files#diff-56c6fa0d10ca0b451d1e53e42a9097a3f5c907a785258cc75b4fe93f01ee0612
WIP for pulling changes to `s3_object_info.`